### ISSUE TITLE
Feature request: n-best paths

### ISSUE THREAD

**Issue by dan-blanchard** [networkx/networkx#793]
As far as I can tell, there's currently no way with networkx to get the n best/shortest paths between two nodes in networkx. I've tried to implement it myself, but it has proved difficult to do efficiently (and in such a way that I could make it allow for ties like all_shortest_paths currently does).

Is there a simple/efficient way to do this I'm just not aware of?

**Comment by Wind4Greg** [comment]
A while ago, for teaching and R&D purposes, I implemented a version of Yen's K-shortest path algorithm in Python/NetworkX.  The code has been posted with the rest of the code for a (data/optical) network course at http://grotto-networking.com/netDesignCourse.html#Code with Sphinx generated documentation at http://grotto-networking.com/files/NetDesignCourse/code_doc/Utilities.html#module-Utilities.YenKShortestPaths.  I recently heard from a biophysics PhD student that has used this code on networks with several thousand nodes and links to find hundreds of shortest paths. 

If there is sufficient interest I can try to modify these to meet NetworkX coding/style/API standards, but would need a bit of guidance from the NetworkX developers.

Cheers
Greg B.

**Comment by aparamon** [comment]
Hello Greg!

It would be great if mine or yours version could be finally integrated into
NextworkX :-)

Regarding your code, I propose that instead of separate methods
findFirstShortestPath, getNextShortestPath, object YenKShortestPaths uses
standard Python iterator protocol (e.g. is generator, see
https://wiki.python.org/moin/Generators).

Best wishes,
Andrey Paramonov

**Comment by Wind4Greg** [comment]
Good idea Andrey, definitely a more Python way to handle the situation.  I'll do an update and post the code with some examples to my website this week.
Cheers
Greg B.

**Comment by Wind4Greg** [comment]
I converted the interface to the K shortest path algorithm (Yen's) to use the iterator pattern, cleaned up all PEP 8 errors (except camel case), Created two different test examples on three different networks and zipped it all to http://www.grotto-networking.com/files/KShortestPaths.zip. 
Cheers
Greg B.

**Comment by jtorrents** [comment]
@Wind4Greg Thanks for looking at this. In a first read of your implementation, I see that you are removing and adding nodes and then searching for shortest paths. We tested this approach in #762 (see this [comment](https://github.com/networkx/networkx/pull/762#issuecomment-12503042), the context manager in the benchmarks is an implementation of the removing and adding nodes/edges approach) and it's quite slower than the filter approach, as in @aparamon nice implementation at #762.

We didn't merge it in the end because the discussion got diverted towards the more general problem of how to efficiently filter nodes and edges in a graph. And whether or not should we change the interface to all shortest paths functions to accept containers for nodes and edges to exclude. We ended up leaving it unfinished, this happens too frequently but time and energy are limited. Sorry for that Andrey!

Anyway, as @chebee7i says, we should revisit the Graph View discussion for 2.0. It would be great to have a GraphView class that could do the filtering (and many more things!). But I think that we should separate it from Yen's algorithm implementation. We can add it now using a private modification of shortest paths functions `bidirectional_shortest_path` and `bidirectional_dijkstra` that accept containers for nodes and edges to exclude during the shortest path search (as in @aparamon implementation). This is what I did in #1405 (although only for nodes), which implements White and Newman algorithm for node independent paths.

If/when we implement the GraphView class, we can rewrite the functions that use private modifications of standard SP algorithms to use it instead. What do you guys think? I'm quite busy, so if someone can step up and put together a PR, that would be great. If not I'll try to do it in the following days.

**Comment by Wind4Greg** [comment]
Hi all, will try to get up to speed on the issues raised by @jtorrents.  Some of those concepts could be useful for some stuff I'm trying to point some of my students at, e.g., k-widest paths.  I make no claims for the speed of my code ;-).  I needed something in Java way back in 2007 or so. Then in 2011 or so needed something in Python.  Didn't have time to get the Epstein algorithm working (couldn't get their C code to compile even), nor time to get the algorithm referenced below to work.

[1]J. Hershberger, M. Maxel, and S. Suri, “Finding the k shortest simple paths: A new algorithm and its implementation,” ACM Trans. Algorithms, vol. 3, no. 4, p. 45, 2007.

Couldn't we (NetworkX) decide on the API for a K-shortest paths algorithm then use Andrey's or my implementation and possibly swap it out in the future if a more efficient implementation gets implemented?

There is a moderately strong need for this functionality in networkX, as I discovered when a PhD candidate in biophysics managed to find my code "hiding" within my website in a network design course zip file.  However his networks were much bigger and he mentioned 30 minute run times with my code (which sounds kind of long to me).

Cheers
Greg

**Comment by aparamon** [comment]
2015-04-06 23:21 GMT+03:00 Greg Bernstein <notifications@github.com>:

> Couldn't we (NetworkX) decide on the API for a K-shortest paths algorithm
> then use Andrey's or my implementation and possibly swap it out in the
> future if a more efficient implementation gets implemented?
>

I believe that the main interface problem is that all_simple_paths name is
already taken :-)

My experience suggests that the user (almost always) naturally expects
simple paths to be ordered by length, and the ability to terminate the
iteration prematurely is invaluable, for serious graphs. It should be the
default interface for getting simple paths of a graph.

Thus, the name yen_simple_paths seems less fortunate (what is yen?).
I think simple_paths is better name in this regard.

Best wishes,
Andrey Paramonov

**Comment by Wind4Greg** [comment]
Hi all, will try to get up to speed on the issues raised by @jtorrents.  Some of those concepts could be useful for some stuff I'm trying to point some of my students at, e.g., k-widest paths.  I make no claims for the speed of my code ;-).  I needed something in Java way back in 2007 or so. Then in 2011 or so needed something in Python.  Didn't have time to get the Epstein algorithm working (couldn't get their C code to compile even), nor time to get the algorithm referenced below to work.

[1]J. Hershberger, M. Maxel, and S. Suri, “Finding the k shortest simple paths: A new algorithm and its implementation,” ACM Trans. Algorithms, vol. 3, no. 4, p. 45, 2007.

Couldn't we (NetworkX) decide on the API for a K-shortest paths algorithm then use Andrey's or my implementation and possibly swap it out in the future if a more efficient implementation gets implemented?

There is a moderately strong need for this functionality in networkX, as I discovered when a PhD candidate in biophysics managed to find my code "hiding" within my website in a network design course zip file.  However his networks were much bigger and he mentioned 30 minute run times with my code (which sounds kind of long to me).

Cheers
Greg

**Comment by jtorrents** [comment]
@Wind4Greg yes, I think we should discuss about an API for k-shortest paths for NetworkX. I'm not sure about what should we do if there are less than k paths between source and target. I'd prefer not to raise an exception (unless there is no path) and just return the ones we found.

I'm preparing a PR with my suggestion above (ie generalize the filer approach of Andrey's implementation in #762 to the weighted case, and use custom private modifications of SP algorithms that accept containers for nodes and edges to exclude). If we go this way, then the k-shortest path function is almost an one liner:

```python
from itertools import islice
def k_shortest_paths(G, source, target, k, weight=None):
        return list(islice(nx.shortest_simple_paths(G, source, target, weight=weight), k))
```

I think this one liner is worth including as a function, but maybe is better if we move this discussion to the PR that I'm about to push.

@aparamon A good reason to keep `all_simple_paths` is its speed and scalability (although with simple paths you get to astronomical number of paths pretty quickly). I did some quick tests, and for small graphs is one order of magnitude faster than `shortest_simple_paths`, for slightly bigger graphs (tens of nodes and density > 0.3)  `all_simple_paths` finished in seconds (reporting millions of paths) and I just ctrl-C `shortest_simple_paths` after running for several minutes in the same graphs.

However I think `shortest_simple_paths` is useful in itself and we should include it, either because the user wants the paths sorted by length, or because she only wants to get the k shortest paths. In this latter case, as long as k is small, it will be quite fast. 

I'll push the PR with my changes now, even if it still needs work, so we can move the discussion there. I'll also add some benchmarks (against `all_simple_paths` and @Wind4Greg implementation).
